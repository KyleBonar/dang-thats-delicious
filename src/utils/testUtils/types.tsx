export interface MockLink {
  text: string;
  to: string;
  target: "_blank" | "_self";
}
