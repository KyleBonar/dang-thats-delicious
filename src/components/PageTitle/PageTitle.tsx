import * as React from "react";

import styled from "../../theme/styled-components";

const StyledH1 = styled.h1`
  font-size: ${props => props.theme.scaleH1};
  margin-bottom: 0.5rem;
`;

interface PageTitleProps {
  children?: string;
  className?: string;
}

const PageTitle: React.SFC<PageTitleProps> = props => {
  return <StyledH1 className={props.className}>{props.children}</StyledH1>;
};

export default PageTitle;
