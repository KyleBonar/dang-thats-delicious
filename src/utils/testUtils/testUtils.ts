import * as React from "react";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";

import { MockCard } from "./types";
import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import { FlashMessageProps } from "../../components/FlashMessage/FlashMessage";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.string,
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.array_of_words(),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = new Array(casual.integer(1, 15))
    .fill(null)
    .map(fakeSauce);
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured: undefined,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const storeConfig = configureStore([]);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const fakeCard = (): MockCard => ({
  title: casual.string,
  description: casual.string,
  to: casual.url,
  showLink: casual.boolean,
  imageLink: casual.url,
  className: casual.string,
  anchorText: casual.string
});

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8) => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(0, MIN_PASSWORD_LENGTH);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword
};
