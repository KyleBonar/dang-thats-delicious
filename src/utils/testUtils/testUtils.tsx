import * as React from "react";
import * as enzyme from "enzyme";
import casual from "casual";
import { ReactWrapper } from "enzyme";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

import { AppState } from "../../redux/configureStore";
import { ISauce, ISaucesState } from "../../redux/sauces/types";
import Flatn from "../Flatn/Flatn";
import { IUserState } from "../../redux/users/types";
import { IErrReturn } from "../Err/Err";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";

export const ITERATION_SIZE = 32;
const REACT_REGEX = /react(\d+)?./i;

// seed it so we get consistent results
casual.seed(777);

const relatedSauce = () => ({
  name: casual.name,
  slug: casual.uuid
});

export const fakeSauce = (): ISauce => ({
  _id: casual.random_element([undefined, casual.integer(1, 500)]),
  _addedToStore: casual.random_element([undefined, casual.unix_time]), // Unix time (in seconds) added to redux store
  _full: casual.random_element([undefined, casual.boolean]), // Whether we have full sauce or partial
  _related: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(relatedSauce)
  ]), // List of related sauces
  isAdminApproved: casual.random_element([undefined, casual.boolean]),
  name: casual.name,
  ingredients: casual.random_element([undefined, casual.string]),
  author: casual.name,
  created: casual.unix_time, // Unix time (in seconds)
  types: casual.random_element([undefined, casual.array_of_words()]),
  maker: casual.name,
  description: casual.string,
  photo: casual.random_element([undefined, casual.url]),
  shu: casual.random_element([
    undefined,
    casual.integer(1, 500),
    casual.integer(1, 500).toString()
  ]),
  reviews: casual.random_element([
    undefined,
    new Array(ITERATION_SIZE).fill(null).map(x => casual.word)
  ]),
  city: casual.random_element([undefined, casual.city]),
  state: casual.random_element([undefined, casual.state]),
  country: casual.random_element([undefined, casual.country]),
  slug: casual.random_element([undefined, casual.uuid])
});

export const generateFakeSauces = (): ISauce[] =>
  new Array(casual.integer(1, 15)).fill(null).map(fakeSauce);

export const fakeSaucesState = (): ISaucesState => {
  const sauces: ISauce[] = generateFakeSauces();
  const { allSlugs, bySlug } = Flatn.saucesArr({ sauces });

  const featured: undefined | string[] = casual.random_element([
    undefined,
    allSlugs.length > 0
      ? allSlugs.filter(slug => {
          if (casual.random < 0.4) return slug;
          return;
        })
      : undefined
  ]);

  return {
    allSlugs,
    bySlug,
    total: allSlugs.length,
    query: undefined,
    saucesWithNewestReviews: undefined,
    newest: undefined,
    featured,
    types: casual.array_of_words(),
    orders: ["Newest", "Name", "Times Reviewed", "Avg Rating"]
  };
};

const fakeUsersState = (): IUserState => ({
  self: casual.random_element([
    undefined,
    {
      token: casual.random_element([undefined, casual.uuid]),
      displayName: casual.random_element([undefined, casual.name]),
      avatarURL: casual.random_element([undefined, casual.url]),
      isAdmin: casual.random_element([undefined, casual.boolean])
    }
  ]),
  byDisplayName: undefined,
  allDisplayNames: undefined
});

export const fakeStore = () => {
  const middlewares = [thunk];
  const storeConfig = configureStore(middlewares);

  const store: AppState = {
    sauces: fakeSaucesState(),
    users: fakeUsersState(),
    reviews: {
      byReviewID: undefined,
      allReviewIDs: undefined
    }
  };

  return storeConfig(store);
};

export const fakeJSXElement = (): JSX.Element => {
  return casual.random_element([
    React.createElement("div", { key: casual.uuid }, casual.string),
    React.createElement("span", { key: casual.uuid }, casual.string)
  ]);
};

const wait = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

function simulateInputChange(
  wrapper: ReactWrapper,
  name: string,
  value: string
) {
  wrapper
    // .find(`input[name='${name}']`)
    .simulate("change", { target: { name, value } });
}

const isClassComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    component.prototype &&
    !!component.prototype.isReactComponent
  );
};

// Ensure compatability with transformed code
const isFunctionComponent = (component): boolean => {
  return (
    typeof component === "function" &&
    String(component).includes("return") &&
    !!String(component).match(REACT_REGEX) &&
    String(component).includes(".createElement")
  );
};

const isElement = (typeElement): boolean => {
  return React.isValidElement(typeElement);
};

const isComponent = (component): boolean => {
  return isClassComponent(component) || isFunctionComponent(component);
};

const isDOMTypeElement = (typeElement): boolean => {
  return isElement(typeElement) && typeof typeElement.type === "string";
};

const generateValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(
    MIN_PASSWORD_LENGTH,
    3 * MIN_PASSWORD_LENGTH
  );

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

const generateInValidPassword = (MIN_PASSWORD_LENGTH: number = 8): string => {
  // generate length between minimum length and 3x the minimum length
  const _randomLength = casual.integer(1, MIN_PASSWORD_LENGTH - 1);

  // turn generated length into random letters
  return new Array(_randomLength).fill(null).map(casual._letter).join("");
};

// Function designed to resemble how a network error would look like.
// This will help the catch(err) only have to handle a single format
const generateErr = (): IErrReturn => {
  return {
    response: {
      data: {
        status: 400,
        msg: casual.string,
        isGood: false
      }
    }
  };
};

// lets us mount up a simple custom hook
const mountReactHook = hook => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.shallow(
      <Component>
        {hookValues => {
          Object.assign(componentHook, hookValues);
          return null;
        }}
      </Component>
    );
  });
  return { componentMount, componentHook };
};

export interface ImountReactHookWithReduxStore {
  componentMount: any;
  componentHook: any;
}
const mountReactHookWithReduxStore = (
  hook,
  store
): ImountReactHookWithReduxStore => {
  const Component = ({ children }) => children(hook());
  const componentHook = {};
  let componentMount;

  act(() => {
    componentMount = enzyme.mount(
      <Provider store={store}>
        <Component>
          {hookValues => {
            Object.assign(componentHook, hookValues);
            return null;
          }}
        </Component>
      </Provider>
    );
  });
  return { componentMount, componentHook };
};

export {
  casual,
  simulateInputChange,
  isDOMTypeElement,
  isComponent,
  wait,
  generateValidPassword,
  generateInValidPassword,
  generateErr,
  mountReactHook,
  mountReactHookWithReduxStore
};
