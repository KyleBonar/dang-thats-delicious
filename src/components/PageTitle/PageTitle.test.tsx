import * as React from "react";
import * as enzyme from "enzyme";

import { casual, ITERATION_SIZE } from "../../utils/testUtils/testUtils";
import PageTitle, { PageTitleProps } from "./PageTitle";

const fakePageTitle = (): PageTitleProps => ({
  children: casual.random_element([undefined, casual.string]),
  className: casual.random_element([undefined, casual.string])
});

const mockPageTitles = new Array(ITERATION_SIZE).fill(null).map(fakePageTitle);

describe("<PageTitle />", () => {
  it("renders", () => {
    mockPageTitles.forEach(mockPageTitle => {
      const wrapper = enzyme.shallow(<PageTitle {...mockPageTitle} />);
      expect(wrapper).toBeTruthy();
    });
  });

  it("matches snapshot", () => {
    mockPageTitles.forEach(mockPageTitle => {
      const wrapper = enzyme.shallow(<PageTitle {...mockPageTitle} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  it("adds className to parent", () => {
    mockPageTitles.forEach(mockPageTitle => {
      if (!mockPageTitle.className) return;
      const wrapper = enzyme.shallow(<PageTitle {...mockPageTitle} />);

      expect(wrapper.find("h1").prop("className")).toContain(
        mockPageTitle.className
      );
    });
  });

  it("renders expected children", () => {
    mockPageTitles.forEach(mockPageTitle => {
      if (!mockPageTitle.children) return;

      const wrapper = enzyme.shallow(<PageTitle {...mockPageTitle} />);

      expect(wrapper.text()).toEqual(mockPageTitle.children);
    });
  });
});
