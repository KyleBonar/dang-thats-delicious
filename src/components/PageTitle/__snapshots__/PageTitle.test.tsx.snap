// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageTitle /> matches snapshot 1`] = `
<h1
  className="fugiat error molestiae laborum est possimus assumenda"
/>
`;

exports[`<PageTitle /> matches snapshot 2`] = `
<h1>
  enim iure facilis ut explicabo iure vitae
</h1>
`;

exports[`<PageTitle /> matches snapshot 3`] = `
<h1>
  soluta non omnis consequatur enim quia autem
</h1>
`;

exports[`<PageTitle /> matches snapshot 4`] = `
<h1>
  laborum eius aut illo dolorum ab aliquam
</h1>
`;

exports[`<PageTitle /> matches snapshot 5`] = `
<h1
  className="cupiditate illum saepe ut assumenda sint quia"
>
  occaecati illum animi vel provident dicta perferendis
</h1>
`;

exports[`<PageTitle /> matches snapshot 6`] = `
<h1
  className="assumenda quasi adipisci mollitia iste itaque omnis"
>
  exercitationem omnis laborum exercitationem est sapiente laudantium
</h1>
`;

exports[`<PageTitle /> matches snapshot 7`] = `
<h1
  className="commodi eaque nisi cum ducimus nemo doloremque"
/>
`;

exports[`<PageTitle /> matches snapshot 8`] = `<h1 />`;

exports[`<PageTitle /> matches snapshot 9`] = `<h1 />`;

exports[`<PageTitle /> matches snapshot 10`] = `
<h1
  className="ullam quae corporis qui aut nobis quo"
>
  explicabo eaque necessitatibus sit suscipit omnis necessitatibus
</h1>
`;

exports[`<PageTitle /> matches snapshot 11`] = `
<h1
  className="occaecati nulla aut sunt est rerum at"
/>
`;

exports[`<PageTitle /> matches snapshot 12`] = `
<h1>
  dolores fugiat quis et voluptas culpa et
</h1>
`;

exports[`<PageTitle /> matches snapshot 13`] = `
<h1
  className="sint officiis qui dignissimos dolore qui voluptatem"
/>
`;

exports[`<PageTitle /> matches snapshot 14`] = `
<h1
  className="quia dolores quisquam aut consequatur sint provident"
>
  nihil iste est unde rem similique et
</h1>
`;

exports[`<PageTitle /> matches snapshot 15`] = `
<h1>
  ullam distinctio autem quam qui facere in
</h1>
`;

exports[`<PageTitle /> matches snapshot 16`] = `<h1 />`;

exports[`<PageTitle /> matches snapshot 17`] = `
<h1>
  fugit in rerum libero odit aut labore
</h1>
`;

exports[`<PageTitle /> matches snapshot 18`] = `
<h1
  className="autem ut ullam et possimus sint fugiat"
/>
`;

exports[`<PageTitle /> matches snapshot 19`] = `
<h1
  className="minus quo sunt possimus earum animi quis"
/>
`;

exports[`<PageTitle /> matches snapshot 20`] = `
<h1
  className="qui sunt nesciunt provident eaque atque quos"
>
  et optio consequatur velit voluptate necessitatibus excepturi
</h1>
`;

exports[`<PageTitle /> matches snapshot 21`] = `<h1 />`;

exports[`<PageTitle /> matches snapshot 22`] = `
<h1
  className="et ipsa consequuntur quidem dolorem quia consequatur"
>
  cupiditate autem consequatur consequatur accusantium molestiae non
</h1>
`;

exports[`<PageTitle /> matches snapshot 23`] = `
<h1
  className="aliquam molestiae occaecati a culpa facilis eligendi"
>
  ab assumenda animi non quis quis nobis
</h1>
`;

exports[`<PageTitle /> matches snapshot 24`] = `<h1 />`;

exports[`<PageTitle /> matches snapshot 25`] = `
<h1
  className="minima temporibus ducimus quia assumenda ut neque"
>
  sint cumque sunt mollitia quibusdam officia eum
</h1>
`;

exports[`<PageTitle /> matches snapshot 26`] = `
<h1>
  est labore sed rerum nesciunt voluptas rem
</h1>
`;

exports[`<PageTitle /> matches snapshot 27`] = `<h1 />`;

exports[`<PageTitle /> matches snapshot 28`] = `
<h1>
  porro ut est sed qui repellat assumenda
</h1>
`;

exports[`<PageTitle /> matches snapshot 29`] = `
<h1
  className="dolor eum velit aperiam voluptas eos quam"
/>
`;

exports[`<PageTitle /> matches snapshot 30`] = `
<h1
  className="et dicta odit vel quisquam qui harum"
/>
`;

exports[`<PageTitle /> matches snapshot 31`] = `
<h1>
  aut esse unde ea voluptatem totam occaecati
</h1>
`;

exports[`<PageTitle /> matches snapshot 32`] = `
<h1
  className="quis ratione eveniet ea totam aut aut"
>
  saepe iure quia culpa suscipit totam libero
</h1>
`;
